src/main.rs

```diff
@@ -42,6 +44,174 @@ impl Shell {
             history: VecDeque::new(),
         }
     }
+
+    pub fn print_prompt(&self) {
+        let prompt_prefix = self.modes.iter().rev().fold(String::new(), |acc, mode| {
+            acc +
+            if mode.value {
+                "+ "
+            } else {
+                "- "
+            }
+        });
+        print!("{}", prompt_prefix);
+
+        let cwd = env::current_dir().ok().map_or("?".to_string(),
+                                                 |ref p| p.to_str().unwrap_or("?").to_string());
+
+        print!("ion:{}# ", cwd);
+        if let Err(message) = stdout().flush() {
+            println!("{}: failed to flush prompt to stdout", message);
+        }
+    }
+
+    fn on_command(&mut self, command_string: &str, commands: &HashMap<&str, Command>) {
+        let max_history: usize = 1000; // TODO temporary, make this configurable
+        if self.history.len() > max_history {
+            self.history.pop_front();
+        }
+        self.history.push_back(command_string.to_string());
+
+        // Show variables
+        if command_string == "$" {
+            for (key, value) in self.variables.iter() {
+                println!("{}={}", key, value);
+            }
+            return;
```

Comment:
Well, you can just add an `else` clause and wrap the rest of the function it it, deleting the `return;` statement.